Jetpack Search: introduce Inline Search

trakos opened the pull request:
Related to https://github.com/Automattic/jetpack-roadmap/issues/2177

## Proposed changes:
* Introduce Inline_Search class
* Introduce `jetpack_search_replace_classic` filter, defaults to false for now

This PR introduces Inline Search, which will eventually replace Classic Search. It uses newer search API (same as Instant Search), but renders search results without popup, similarly to Classic Search. For now, the feature is behind a feature flag, so this PR won't introduce any changes unless you opt-in using `jetpack_search_replace_classic` filter.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pgbpWj-4q-p2

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:

37262-pb/#plain

To avoid cache when testing I'd usually just add a space to my search query text.

trakos committed:
Jetpack Search: introduce Smart Inline Search

trakos committed:
Jetpack Search: introduce Smart Inline Search

sotirispl reviewed (dismissed):
Tested and works as described.

trakos committed:
Rename Smart_Inline_Search to Inline_Search, keep full search response, overwrite more code from classic

trakos commented:
I've updated this PR with the following changes:

- Renamed Smart_Inline_Search to Inline_Search (as disussed during a team meeting).
- Changed the search_result variable to keep the full response, so that we can use additional parameters like corrected_query for upcoming features.
- This also required moving more code from Classic Search to Inline Search class, since I needed to tweak the way it accesses search_result value. We would need to do this eventually anyway to change how some of the filters are called.

gibrown commented on projects/packages/search/src/smart-inline-search/class-inline-search.php:
```diff
@@ -0,0 +1,409 @@
+/**
+ * Inline Search class
+ */
+class Inline_Search extends Classic_Search {
```
should we change the "smart-inline-search" directory for this file also?

trakos replied:
Good point, I missed that when renaming the class. I've renamed the directory to "inline-search" now.

gibrown reviewed (commented):
Direction makes sense to me and looks good. Only saw the directory naming problem which is maybe a blocker

trakos committed:
Fix inline-search directory name

sotirispl commented on projects/packages/search/src/widgets/class-search-widget.php:
```diff
@@ -324,12 +324,17 @@ public function widget_non_instant( $args, $instance ) {
 		// Search instance must have been initialized before widget render.
-		if ( is_search() && Classic_Search::instance() ) {
+		if ( is_search() ) {
+			if ( Inline_Search::should_replace_classic_search() ) {
```
I am not sure if it is feasible, but we could move this block to a function in the `class-initializer.php` since it is used there and here for the widget.

trakos replied:
Makes sense, I've introduced `get_instance_maybe_fallback_to_classic` method to address this.

sotirispl reviewed (dismissed):
I left a comment with a possible improvement point in the code, but not that significant to block this pull request. LGTM!

trakos committed:
Introduce get_instance_maybe_fallback_to_classic method

trakos commented:
Thank you! I've updated the PR with the suggested method.
